### PR TITLE
Finish tsm file exporter

### PR DESCRIPTION
**Pull request by corylanou**
Add ability to export TSM files to line protocol to be imported back into another database.

- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated
- [x] Update docs issue created
- [x] Update README.md

### CODE REVIEW

**Comment by beckettsean** [comment]
@gunnaraasen

**Comment by dgnorton** [comment]
Doesn't this functionality exist already in https://github.com/influxdata/influxdb/blob/master/cmd/influx_inspect/export.go?

File: cmd/influx_inspect/export.go [hunk 1]
```diff
@@ -1,92 +1,143 @@
...
+			file, err := os.OpenFile(f, os.O_RDONLY, 0600)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			defer file.Close()
+			reader, err := tsm1.NewTSMReader(file)
+			if err != nil {
+				return err
+			}
+			defer reader.Close()
```
Comment (mark-rushakoff) [inline cmd/influx_inspect/export.go]: These deferred `Close`s won't execute until the end of `writeFiles` - maybe wrap the inner loop in an anonymous function?

File: cmd/influx_inspect/export.go [hunk 2]
```diff
@@ -1,92 +1,143 @@
...
+		log.Fatal(err)
+	}
+	return nil
+}
 
-			w.Flush()
+func (c *cmdExport) writeFiles() error {
+	// open our output file and create an output buffer
+	var w io.WriteCloser
+	w, _ = os.Create(c.out)
```
Comment (mark-rushakoff) [inline cmd/influx_inspect/export.go]: It's easy to make a mistake in specifying the output path, so this error should not be ignored.

Comment (corylanou) [inline cmd/influx_inspect/export.go]: ah, correct, I meant to update that code.  Good catch.

File: cmd/influx_inspect/export.go [hunk 3]
```diff
@@ -1,92 +1,143 @@
...
+			if err != nil {
+				return err
+			}
+			defer reader.Close()
+
+			for i := 0; i < reader.KeyCount(); i++ {
+				var pairs string
+				key, typ := reader.KeyAt(i)
+				values, _ := reader.ReadAll(key)
+				split := strings.Split(key, "#!~#")
```
Comment (jwilder) [inline cmd/influx_inspect/export.go]: There is a `tsdb.seriesAndFieldFromCompositeKey` that does this.  Might be better to export and use that instead.

File: cmd/influx_inspect/export.go [hunk 4]
```diff
@@ -1,92 +1,143 @@
...
+				for _, value := range values {
+					switch typ {
+					case tsm1.BlockFloat64:
+						pairs = field + "=" + fmt.Sprintf("%v", value.Value())
+					case tsm1.BlockInteger:
+						pairs = field + "=" + fmt.Sprintf("%vi", value.Value())
+					case tsm1.BlockBoolean:
+						pairs = field + "=" + fmt.Sprintf("%v", value.Value())
+					case tsm1.BlockString:
+						pairs = field + "=" + fmt.Sprintf("\"%v\"", value.Value())
```
Comment (jwilder) [inline cmd/influx_inspect/export.go]: Needs to be passed through `models.escapeStringField`.

File: cmd/influx_inspect/export.go [hunk 5]
```diff
@@ -1,92 +1,143 @@
...
+				return fmt.Errorf("%v", err)
+			}
+			defer file.Close()
+			reader, err := tsm1.NewTSMReader(file)
+			if err != nil {
+				return err
+			}
+			defer reader.Close()
+
+			for i := 0; i < reader.KeyCount(); i++ {
```
Comment (jwilder) [inline cmd/influx_inspect/export.go]: This is the best way.  It arrives as a `Point`, but it's not stored or retrieved in that format ever again.

File: cmd/influx_inspect/main.go [hunk 6]
```diff
@@ -101,9 +102,14 @@ func main() {
 		}
 		cmdVerify(path)
 	case "export":
-		var path string
+		var path, out, database, rp string
+		var compress bool
 		fs := flag.NewFlagSet("export", flag.ExitOnError)
 		fs.StringVar(&path, "dir", os.Getenv("HOME")+"/.influxdb", "Root storage path. [$HOME/.influxdb]")
+		fs.StringVar(&out, "out", os.Getenv("HOME")+"/.influxdb/export", "Destination file to export to")
+		fs.StringVar(&database, "database", "", "Optional: the database to export")
```
Comment (jwilder) [inline cmd/influx_inspect/main.go]: Shorted to `db` to match `rp`?

File: cmd/influx_inspect/export.go [hunk 7]
```diff
@@ -1,92 +1,143 @@
...
+		fmt.Fprintf(w, "# CONTEXT-RETENTION-POLICY:%s\n", keys[1])
+		for _, f := range files {
+			file, err := os.OpenFile(f, os.O_RDONLY, 0600)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			defer file.Close()
+			reader, err := tsm1.NewTSMReader(file)
+			if err != nil {
+				return err
```
Comment (jwilder) [inline cmd/influx_inspect/export.go]: Log (to stderr) and skip the file if it fails to load?

File: cmd/influx_inspect/export.go [hunk 8]
```diff
@@ -1,92 +1,143 @@
...
-			w.Flush()
+func (c *cmdExport) writeFiles() error {
+	// open our output file and create an output buffer
+	var w io.WriteCloser
+	w, _ = os.Create(c.out)
+	defer w.Close()
+	if c.compress {
+		w = gzip.NewWriter(w)
+	}
+	fmt.Fprintln(w, "# DML")
```
Comment (jwilder) [inline cmd/influx_inspect/export.go]: We know the DB and RP from path info, seems like we add those create statements.  They'll be a no-op if the same defaults exist.

File: cmd/influx_inspect/export.go (hunk 5 above)
Comment (jwilder) [inline cmd/influx_inspect/export.go]: Actually, the fastest way to walk the file is to use a `BlockIterator` and then decode each block based on the type.  `reader.ReadAll` via keys works, but will be slower and allocated a lot of memory.

File: cmd/influx_inspect/export.go (hunk 8 above)
Comment (corylanou) [inline cmd/influx_inspect/export.go]: We don't know the policy for the RP though...  I think it will err out if it doesn't match.

Comment (jwilder) [inline cmd/influx_inspect/export.go]: It will error out if you try to create an rp with the same name, but exists differently.

**Comment by corylanou** [comment]
@jwilder comments addressed.  Ready for review.

File: cmd/influx_inspect/export.go [hunk 9]
```diff
@@ -1,92 +1,162 @@
...
+				for i := 0; i < reader.KeyCount(); i++ {
+					var pairs string
+					key, typ := reader.KeyAt(i)
+					values, _ := reader.ReadAll(key)
+					measurement, field := tsm1.SeriesAndFieldFromCompositeKey(key)
+
+					for _, value := range values {
+						switch typ {
+						case tsm1.BlockFloat64:
+							pairs = field + "=" + fmt.Sprintf("%v", value.Value())
```
Comment (jwilder) [inline cmd/influx_inspect/export.go]: Will this format `float64(0)` as `0.0`?  Floats should include a decimal to ensure their types won't switch to an int inadvertently.

File: cmd/influx_inspect/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+#### Sample Commands
+
+Export entire database and compress output:
+```
+influx_inspect export --compress
+```
+
+Export specific retention policy:
+```
+influx_inspect export --database mydb --rp autogen
```
Comment (jwilder) [inline cmd/influx_inspect/README.md]: `-db`

File: cmd/influx_inspect/README.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
+# `influx_inspect`
+
+## Ways to run
+
+### `influx_inspect`
+Will print usage for the tool.
+
+### `influx_inspect info`
```
Comment (jwilder) [inline cmd/influx_inspect/README.md]: I'm replacing this command w/ `report` in #7047

**Comment by jwilder** [comment]
Two small things, but looks good! 👍

File: cmd/influx_inspect/README.md (hunk 11 above)
Comment (corylanou) [inline cmd/influx_inspect/README.md]: If I merge first it means you have to update... ok, fine, I'll do it :-)

**Comment by retorquere** [comment]
Is this part of the 1.0 beta?

**Comment by jwilder** [comment]
@retorquere It's in the current nightlies.

**Comment by retorquere** [comment]
Ok, no issue. Is this compatible with 0.13 and could I compile it separately? Alternatively, are the nightlies usable enough to be mostly-available, and is there a clean upgrade path between 0.13?

Is there a timeline for 1.0?

**Comment by kostasb** [comment]
@retorquere The upgrade bath from 0.13 is straight-forward, installing the new package over the existing 0.13 will be sufficient. Please review the Changelog before upgrading in case you want to modify the new default configuration values (particularly max-series-per-database).

**Comment by retorquere** [comment]
Thanks for the feedback, and I'm looking at that parameter now, but this is about the number of series, not taking into account the cardinality of the tags? Not that it would affect me I think; I have 3 series, each with one tag with cardinality 3, but with 70 fields.

**Pull request merged**